hugegraph-core/src/main/java/com/baidu/hugegraph/task/StandardTaskScheduler.java

```diff
@@ -248,65 +271,109 @@ private void unlistenChanges() {
     @Override
     public synchronized <V> void cancel(HugeTask<V> task) {
         E.checkArgumentNotNull(task, "Task can't be null");
-        if (!this.serverManager().master()) {
+        this.checkOnMasterNode("cancel");
+
+        if (task.completed() || task.cancelling()) {
             return;
         }
-        if (!task.completed()) {
-            // The task scheduled to workers, waiting for worker cancel
-            task.status(TaskStatus.CANCELLING);
+
+        LOG.info("Cancel task '{}' in status {}", task.id(), task.status());
+
+        if (task.server() == null) {
+            // The task not scheduled to workers, set canceled immediately
+            assert task.status().code() < TaskStatus.QUEUED.code();
+            if (task.status(TaskStatus.CANCELLED)) {
+                this.save(task);
+                return;
+            }
+        } else if (task.status(TaskStatus.CANCELLING)) {
+            // The task scheduled to workers, let the worker node to cancel
             this.save(task);
-            this.remove(task.id());
+            assert task.server() != null : task;
+            assert this.serverManager().master();
+            if (!task.server().equals(this.serverManager().selfServerId())) {
+                /*
+                 * Remove task from memory if it's running on worker node,
+                 * but keep task in memory if it's running on master node.
+                 * cancel-scheduling will read task from backend store, if
+                 * removed this instance from memory, there will be two task
+                 * instances with same id, and can't cancel the real task that
+                 * is running but removed from memory.
+                 */
+                this.remove(task);
+            }
```

Comment:
else means task is on master, cancel it directly seems more efficient

Reply:
We assume cancel is not a frequent operation, so it is not optimized